File: E-DES.c
/**
 * @file E-DES.c
 * @brief Implementation of the E-DES algorithm
 * @author Fábio Santos 1188351
 * @author Rodrigo Marques 118587
 * @date 05/11/2023
*/

#include "E-DES.h"

// ============================================================
// ======================== GERAL =============================
// ============================================================

void sha256(char *string, uint8_t *hash) {
    EVP_MD_CTX *mdctx;
    const EVP_MD *md;

    md = EVP_sha256();
    mdctx = EVP_MD_CTX_new();

    EVP_DigestInit_ex(mdctx, md, NULL);
    EVP_DigestUpdate(mdctx, string, strlen(string));
    EVP_DigestFinal_ex(mdctx, hash, NULL);
    EVP_MD_CTX_free(mdctx);
}

void sha256_uint8(uint8_t *input, size_t input_len, uint8_t *output) {
    EVP_MD_CTX *mdctx;
    const EVP_MD *md;

    md = EVP_sha256();
    mdctx = EVP_MD_CTX_new();

    if (mdctx == NULL) {
        printf("Error in sha256\n");
        return;
    }

    if (EVP_DigestInit_ex(mdctx, md, NULL) != 1 ||
        EVP_DigestUpdate(mdctx, input, input_len) != 1 ||
        EVP_DigestFinal_ex(mdctx, output, NULL) != 1) {
        EVP_MD_CTX_free(mdctx);
        printf("Error in sha256\n");
        return; // Hashing failed
    }

    EVP_MD_CTX_free(mdctx);
}

uint8_t *read_msg_bytes(uint64_t *bytes_read) {
    uint8_t aux;
    uint8_t *input_bytes = NULL;

    (*bytes_read) = 0;
    while (fread(&aux, sizeof(uint8_t), 1, stdin) == 1) {
        input_bytes = realloc(input_bytes, (++(*bytes_read)) * sizeof(uint8_t));
        if (!input_bytes) {
            fprintf(stderr, "Failed to realloc memory!\n");
            return NULL;
        }
        input_bytes[(*bytes_read) - 1] = aux;
    }
    return input_bytes;
}

void bytes_to_char(uint8_t *input, char *output, int len) {
    for (int i = 0; i < len; i++) {
        output[i] = (char)input[i];
    }
}

void add_padding(uint8_t *input, uint64_t *bytes_read) {

    // Add padding
    int block_size = 8;
    if ((*bytes_read) % block_size != 0) {
        // Realloc the input
        int padding_size = block_size - ((*bytes_read) % block_size);
        input = (uint8_t *)realloc(input, ((*bytes_read) + padding_size) * sizeof(uint8_t));
        // Add padding to the rest of the block
        for (int i = (*bytes_read); i < (*bytes_read) + padding_size; i++) {
            input[i] = (uint8_t)padding_size;
        }
        (*bytes_read) += padding_size;
    } else {
        // Realloc the input
        input = (uint8_t *)realloc(input, ((*bytes_read) + block_size) * sizeof(uint8_t));
        // Add another block of padding
        for (int i = (*bytes_read); i < (*bytes_read) + block_size; i++) {
            input[i] = (uint8_t)block_size;
        }
        (*bytes_read) += block_size;
    }
}

void remove_padding(uint8_t *input, uint64_t *len) {
    
    // int padding_size = input[len - 1] == 0 ? input[len - 2] : input[len - 1];
    uint8_t padding_size = input[(*len) - 1];
    (*len) -= padding_size;
    input[(*len) - padding_size] = '\0';
}


void read_msg(char **msg) {
    int block_size = 8;
    size_t bytes_read = 0;

    // Read size bytes from stdin
    char *buffer = NULL;
    char c;

    while (1) {
        c = getchar();
        if (c == EOF || c == '\n') {
            break;
        }
        buffer = (char *)realloc(buffer, (bytes_read + 1) * sizeof(char));
        buffer[bytes_read] = c;
        bytes_read++;
    }

    int padding_size = block_size - (bytes_read % block_size);
    // Copy the buffer to the input
    *msg = (char *)malloc((bytes_read + 1) * sizeof(char));
    (*msg)[0] = '\0';
    strcat(*msg, buffer);
    free(buffer);
}

void create_SBoxes(SBox *SBoxes, char *password) {

    // Inicialize the SBoxes
    for (int i = 0; i < 16; i++) {
        for (uint8_t j = 0x00; j < 0xff; j++) {
            SBoxes[i].s[j] = j;
        }
        SBoxes[i].s[0xff] = 0xff;
    }
    
    
    // Shuffle the SBoxes using the password
    uint8_t *hash = (uint8_t *)malloc(SHA256_DIGEST_LENGTH * sizeof(uint8_t));
    if (hash == NULL) {
        printf("Error allocating memory\n");
        return; // Memory allocation failed
    }

    sha256(password, hash);

    for (int i = 0; i < 16; i++) {

        // Shuffle the SBox
        for (int j = 0; j < 256; j++) {
            uint8_t value = hash[j % 32];

            uint8_t aux = SBoxes[i].s[j];
            SBoxes[i].s[j] = SBoxes[i].s[value];
            SBoxes[i].s[value] = aux;
        }

        // Get the next hash (hash of previous hash)
        sha256_uint8(hash, SHA256_DIGEST_LENGTH, hash);
    }
}


void f_SBox(SBox SBox, uint8_t input[4], uint8_t output[4]) {
    uint8_t index = input[3];
    output[0] = SBox.s[index];

    index = (index + input[2]);
    output[1] = SBox.s[index];

    index = (index + input[1]);
    output[2] = SBox.s[index];

    index = (index + input[0]);
    output[3] = SBox.s[index];
}

// ============================================================
// ======================== ENCRYPT ===========================
// ============================================================


void feistel_networks_block(SBox *SBoxes, uint8_t block[8], uint8_t output[8]) {
    // uint8_t block[8]correponde a um bloco de 64 bits / 8 bytes; cada bloco vai ter 4 carateres

    // Split the block in 2 equal parts
    uint8_t block_left[4];
    uint8_t block_right[4];
    for (int i = 0; i < 8; i++) {
        if (i < 4) {
            block_left[i] = block[i];
        } else {
            block_right[i - 4] = block[i];
        }
    }

    for (int i = 0; i < 16; i++) {
        // Create the next blocks
        uint8_t block_left_next[4];
        uint8_t block_right_next[4];

        // The next block left is the current block right
        for (int j = 0; j < 4; j++) {
            block_left_next[j] = block_right[j];
        }

        // The next block right is the current block left XOR result of the current block right with correspondent S-Box
        f_SBox(SBoxes[i], block_right, block_right_next);
        for (int j = 0; j < 4; j++) {
            block_right_next[j] = block_right_next[j] ^ block_left[j];
        }

        // Update the blocks
        for (int j = 0; j < 4; j++) {
            block_left[j] = block_left_next[j];
            block_right[j] = block_right_next[j];
        }
    }

    // Merge the blocks
    for (int i = 0; i < 8; i++) {
        if (i < 4) {
            output[i] = block_left[i];
        } else {
            output[i] = block_right[i - 4];
        }
    }
}


void feistel_networks(SBox *SBoxes, uint8_t *input, uint8_t *output, uint64_t input_len) {
    uint64_t num_blocks = input_len / 8;
    for (uint64_t i = 0; i < num_blocks; i++) {
        feistel_networks_block(SBoxes, &input[i * 8], &output[i * 8]);
    }
}

// ============================================================
// ======================== DECRYPT ===========================
// ============================================================

void feistel_networks_block_decrypt(SBox *SBoxes, uint8_t block[8], uint8_t output[8]) {

    // Split the block in 2 equal parts
    uint8_t block_left[4];
    uint8_t block_right[4];
    for (int i = 0; i < 8; i++) {
        // Aqui troca em relacao ao encrypt
        if (i < 4) {
            block_right[i] = block[i];
        } else {
            block_left[i - 4] = block[i];
        }
    }

    for (int i = 15; i >= 0; i--) {
        // Create the previous blocks
        uint8_t block_left_previous[4];
        uint8_t block_right_previous[4];

        // The previous block left is the current block right
        for (int j = 0; j < 4; j++) {
            block_left_previous[j] = block_right[j];
        }

        // The previous block right is the current block left XOR result of the current block right with correspondent S-Box
        f_SBox(SBoxes[i], block_right, block_right_previous);
        for (int j = 0; j < 4; j++) {
            block_right_previous[j] = block_right_previous[j] ^ block_left[j];
        }

        // Update the blocks
        for (int j = 0; j < 4; j++) {
            block_left[j] = block_left_previous[j];
            block_right[j] = block_right_previous[j];
        }
    }

    // Merge the blocks
    for (int i = 0; i < 8; i++) {
        // Aqui troca em relacao ao encrypt
        if (i < 4) {
            output[i] = block_right[i];
        } else {
            output[i] = block_left[i - 4];
        }
    }
}

void feistel_networks_decrypt(SBox *SBoxes, uint8_t *input, uint8_t *output, uint64_t input_len) {
    uint64_t num_blocks = input_len / 8;
    for (uint64_t i = 0; i < num_blocks; i++) {
        feistel_networks_block_decrypt(SBoxes, &input[i * 8], &output[i * 8]);
    }
}

File: EDES.py
"""!
@file EDES.py
@brief This file contains the implementation of the EDES algorithm.
@author Fábio Santos 118351
@author Rodrigo Marques 118587
@date 05/01/2023
"""

import sys
import hashlib

PW_LEN = 32

# ============================================================
# ======================== GERAL =============================
# ============================================================
#make a documentation for this function


def sha256(input):
    '''!
    @brief This function returns the sha256 of the input.
    @param input The input to be hashed.
    '''
    sha256 = hashlib.sha256()
    sha256.update(input.encode("utf-8"))
    sha256_bytes = sha256.digest()
    sha256_bytearray = bytearray(sha256_bytes)

    return sha256_bytearray


def sha256_bytearray(byte_array):
    '''!
    @brief This function returns the sha256 of the input.
    @param input The input to be hashed.
    '''

    sha256 = hashlib.sha256()
    sha256.update(byte_array)
    sha256_bytes = sha256.digest()
    sha256_bytearray = bytearray(sha256_bytes)

    return sha256_bytearray


def read_to_bytearray():
    '''!
    @brief This function reads the input from stdin into a bytearray.
    @return The bytearray containing the input and the number of bytes read.
    '''

    try:
        # Read input from stdin into a bytearray
        # data = bytearray(sys.stdin.buffer.read())
        data = bytearray(sys.stdin.buffer.read())
        bytes_read = len(data)

        return data, bytes_read
    except Exception as e:
        print("An error occurred:", e)
        return None, 0


def add_padding(input, bytes_read):
    '''!
    @brief This function adds padding to the input.
    @param input The input to be padded.
    @param bytes_read The number of bytes read from stdin.
    @return The padded input.
    '''

    if (bytes_read % 8 != 0):
        padding_size = 8 - (bytes_read % 8)
        for _ in range(padding_size):
            input.append(padding_size)
    else:
        for _ in range(8):
            input.append(0x08)
    return input


def remove_padding(input):
    '''!
    @brief This function removes the padding from the input.
    @param input The input to be unpadded.
    @return The unpadded input.
    '''

    padding_size = int(input[-1])
    return input[:-padding_size]


def create_SBoxes(password):
    '''!
    @brief This function creates the SBoxes.
    @param password The password to be used to create the SBoxes.
    @return The SBoxes.
    '''

    # Init the SBoxes
    SBoxes = []
    for _ in range(16):
        SBoxes.append(bytearray([j for j in range(256)]))

    hash = sha256(password)

    # Shuffle the SBoxes
    for i in range(16):
        for j in range(256):
            value = hash[j % 32]
            aux = SBoxes[i][j]
            SBoxes[i][j] = SBoxes[i][value]
            SBoxes[i][value] = aux
        hash = sha256_bytearray(hash)

    return SBoxes


def f_SBox(SBox, input):
    '''!
    @brief This function applies the SBox to the input.
    @param SBox The SBox to be used.
    @param input The input to be used.
    @return The output of the SBox.
    '''

    index = input[3] % 256
    output = [SBox[index]]

    index = (index + input[2]) % 256
    output.append(SBox[index])

    index = (index + input[1]) % 256
    output.append(SBox[index])

    index = (index + input[0]) % 256
    output.append(SBox[index])

    return output


# ============================================================
# ======================== ENCRYPT ===========================
# ============================================================


def feistel_networks_block(SBoxes, block):
    '''!
    @brief This function applies the feistel networks to a block.
    @param SBoxes The SBoxes to be used.
    @param block The block to be used.
    @return ciphered block.
    '''

    output = bytearray(8)
    block_left = block[:4]
    block_right = block[4:]

    for i in range(16):
        block_left_next = block_right[:]
        block_right_next = bytearray(4)

        block_right_next = f_SBox(SBoxes[i], block_right)
        for j in range(4):
            block_right_next[j] = block_right_next[j] ^ block_left[j]

        block_left = block_left_next[:]
        block_right = block_right_next[:]

    output[:4] = block_left[:]
    output[4:] = block_right[:]

    return output


def feistel_networks(Sboxes, input):
    '''!
    @brief This function applies the feistel networks to the input.
    @param SBoxes The SBoxes to be used.
    @param input The input to be used.
    @return ciphered input.
    '''

    num_blocks = int(len(input) / 8)
    output = bytearray(len(input))
    for i in range(num_blocks):
        block = input[i*8: i*8+8]
        output[i*8: i*8+8] = feistel_networks_block(Sboxes, block)

    return output


# ============================================================
# ======================== DECRYPT ===========================
# ============================================================


def feistel_networks_block_decrypt(SBoxes, block):
    '''!
    @brief This function applies the feistel networks to a block to decrypt it
    @param SBoxes The SBoxes to be used.
    @param block The block to be used.
    @return deciphered block.
    '''

    output = bytearray(8)
    block_left = block[4:]
    block_right = block[:4]

    for i in range(15, -1, -1):
        block_left_previous = block_right[:]
        block_right_previous = bytearray(4)

        block_right_previous = f_SBox(SBoxes[i], block_right)
        for j in range(4):
            block_right_previous[j] = block_right_previous[j] ^ block_left[j]

        block_left = block_left_previous[:]
        block_right = block_right_previous[:]

    output[:4] = block_right[:]
    output[4:] = block_left[:]

    return output


def feistel_networks_decrypt(Sboxes, input):
    '''!
    @brief This function applies the feistel networks to the input to decrypt it.
    @param SBoxes The SBoxes to be used.
    @param input The input to be used.
    @return deciphered input.
    '''

    num_blocks = int(len(input) / 8)
    output = bytearray(len(input))
    for i in range(num_blocks):
        block = input[i*8: i*8+8]
        output[i*8: i*8+8] = feistel_networks_block_decrypt(Sboxes, block)

    return output
